custom-services/custom-service.md

```diff
@@ -0,0 +1,57 @@
+
+# Proposed Solution
+
+### Add enable mobile button / link
+To allow a developer to expose a service already provisioned from the service catalog, we could modify the UI view in OpenShift to have a call
+to action (something like) "enable mobile" that would show up on the overview screen inside the ServiceInstance box if mobile was enabled within the cluster and there was at least one mobile client in the namespace. When clicked they would be presented with a modal form that would allow them to specify needed configuration details (url, API Keys, any other key value pairs needed in order to consume the service). The key value pairs would be dynamic (as in they could add as many as needed) there is already options for this type of field in the OpenShift UI (used to specify env vars).
+When saved we would create the required objects (ie the configmap and a secret if needed) in the right format and with the correct labels to allow it to be consumed from the mobile CLI.
+
+**Considerations**
+
+Do we add this button to all services, it doesn't make sense to add it to something like MongoDB for example.
+
+## Create Custom Service Connector APB
+
+To allow mobile developers to expose existing custom logic contained within running services that may or may not be running on an OpenShift cluster, I propose we look at creating a custom service connector APB.
+
+All of our mobile enabled services are provisioned via the catalog. I believe we should follow the same pattern for allowing mobile developers to expose custom services to the mobile application clients. This would give us the following advantages:
+
+- It would create a ServiceInstance that we can then leverage as part of the UI work, ensuring we have a single Object to deal with for all service interactions in the UI and CLI.
+
+- It would provide a familiar flow to developers of going to the catalog and setting up a service. 
+
+- It would give us a simple way to allow setting up a custom service via the CLI
+
+- It would ensure we have control over what and how the service is setup so that we can consume it both from the UI, CLI.
+
+### How it would work
+
+When you provisioned a custom service connector from the catalog, you would be shown a form that would have some required fields such as the host and a name for the service. Then it would accept dynamic key value pairs (in a similar fashion to what is done now for env vars). This would allow a developer to specify important configuration information needed to consume the service such as an API Key etc. As part of the provision we would take this information and create the required objects (likely just a configmap and perhaps a secret) and label them accordingly. As there would be a ServiceInstance created, the developer would also be able to add and exclude clients from the this service and deprovision it. To retrieve the configuration the developer would continue to use the UI screen or the CLI in exactly the same manner as before.
```

Comment:
This looks to be saying the same thing as some of the sentences in the `Add enable mobile button / link` section. Maybe this section is not required if it's already detailed?

Reply:
They are doing similar things but they are different flows. 
- already provisioned a service with a service instance so just show a modal
- older or external service so go through catalog and use custom service APB that will then setup the service instance. 
Maybe I need to make this clearer? I outlined the use cases early on

Reply:
I think it's OK as is. Thanks for clearing it up